.dotnet/OpenAI.Microsoft.Generator.CSharp.ClientModel.Plugin/OpenAILibraryVisitor.cs

```diff
@@ -0,0 +1,174 @@
+using static Microsoft.Generator.CSharp.Snippets.Snippet;
+
+namespace Microsoft.Generator.CSharp.ClientModel.OpenAILibraryPlugin
+{
+    public class OpenAILibraryVisitor : ScmLibraryVisitor
+    {
+        private const string RawDataPropertyName = "SerializedAdditionalRawData";
+        private const string AdditionalPropertiesFieldName = "_additionalBinaryDataProperties";
+        private const string SentinelValueFieldName = "_sentinelValue";
+        private const string WritePropertyNameMethodCall = "WritePropertyName(\"";
+
+        private static readonly VariableExpression AdditionalBinaryDataExpression =
+            new VariableExpression(typeof(IDictionary<string, BinaryData>), AdditionalPropertiesFieldName);
+        protected override TypeProvider Visit(TypeProvider type)
+        {
+            if (type is ModelProvider { BaseModelProvider: null })
+            {
+                var additionalPropertiesField = type.Fields.Single(f => f.Name == AdditionalPropertiesFieldName);
+                var properties = new List<PropertyProvider>(type.Properties)
+                {
+                    new PropertyProvider($"", MethodSignatureModifiers.Internal,
+                        typeof(IDictionary<string, BinaryData>), RawDataPropertyName,
+                        new ExpressionPropertyBody(
+                            additionalPropertiesField,
+                            additionalPropertiesField.Assign(Value)), type)
+                };
+                type.Update(properties: properties);
+            }
+            else if (type.Name == "ModelSerializationExtensions")
+            {
+                var sentinelValueField = new FieldProvider(
+                    FieldModifiers.Private | FieldModifiers.Static | FieldModifiers.ReadOnly,
+                    typeof(BinaryData),
+                    SentinelValueFieldName,
+                    type,
+                    null,
+                    BinaryDataSnippets.FromBytes(LiteralU8("\"__EMPTY__\"").Invoke("ToArray")));
+                var fields = new List<FieldProvider>(type.Fields)
+                {
+                    sentinelValueField
+                };
+
+                var valueParameter = new ParameterProvider("value", $"", typeof(BinaryData));
+                var methods = new List<MethodProvider>(type.Methods)
+                {
+                    new MethodProvider(
+                        new MethodSignature(
+                            "IsSentinelValue",
+                            $"",
+                            MethodSignatureModifiers.Internal | MethodSignatureModifiers.Static,
+                            typeof(bool),
+                            $"",
+                            [valueParameter]),
+                        new[]
+                        {
+                            Declare("sentinelSpan", typeof(ReadOnlySpan<byte>), sentinelValueField.Invoke("ToMemory", []).Property("Span"), out var sentinelVariable),
```

Comment:
could we use [this snippet](https://github.com/microsoft/typespec/blob/main/packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Snippets/BinaryDataSnippets.cs#L28) here ?